jax/_src/source_info_util.py

```diff
@@ -33,38 +33,100 @@
 def register_exclusion(path):
   _exclude_paths.append(path)
 
-def user_frames(source_info: Optional[Traceback]) -> Iterator[Frame]:
+class Scope(NamedTuple):
+  name: str
+
+  def wrap(self, stack: Tuple[str, ...]) -> Tuple[str, ...]:
+    return (self.name, *stack)
+
+class Transform(NamedTuple):
+  name: str
+
+  def wrap(self, stack: Tuple[str, ...]) -> Tuple[str, ...]:
+    return tuple(map(lambda x: f'{self.name}({x})', stack))
+
+class NameStack(NamedTuple):
+  stack: Tuple[Union[Scope, Transform], ...] = ()
+
+  def extend(self, name: Union[Tuple[str, ...], str]) -> 'NameStack':
+    if not isinstance(name, tuple):
+      name = (name,)
+    scopes = tuple(map(Scope, name))
+    return NameStack(self.stack + scopes)
```

Comment:
Some golf: instead of

```python
    scopes = tuple(map(Scope, name))
    return NameStack(self.stack + scopes)
```

we could write

```python
return NameStack((*self.stack, *map(Scope, name)))
```